File: hackerrank/practice/matrix-rotation-algo.py
from collections import deque

def lin(a, m, n):
	"""This function will take a matrix of m*n dimensions, and will give a two dimensional array, which is linearized in its layers."""
	nol = min(n, m)/2 #no of layers in the matrix
	temp = deque()
	startx = 0
	starty = 0
	while startx < nol:
		temp2 = deque()
		x = startx
		y = starty
		while x < m-startx:
			temp2.append(a[x][y])
			# print "stuck in here>", x
			x += 1
		x -= 1
		temp2.pop()
		while y < n-starty:
			temp2.append(a[x][y])
			# print "Entered1"
			y += 1
		y -= 1
		temp2.pop()
		while x >= startx:
			# print "Entered2"
			temp2.append(a[x][y])
			x -= 1
		x += 1
		temp2.pop()
		while y >= starty:
			# print "Entered3"
			temp2.append(a[x][y])
			y -= 1
		y += 1
		temp2.pop()
		startx += 1
		starty += 1
		temp.append(temp2)
	return temp

def rotate(a, k):
	"""This function takes a queue and rotates it k times, and returns it."""
	k %= len(a)
	while k:
		val = a.pop()
		a.appendleft(val)
		k -= 1
	return a

def form(a, m, n):
	temp = [[0 for i in xrange(n)] for j in xrange(m)]
	startx = 0
	starty = 0
	nol = min(m, n)/2
	while startx < nol:
		x = startx
		y = starty
		i = 0
		# print "iteration: ",startx+1
		try:
			while x < m-startx:
				# print "x",x,"y",y,"val:",a[startx][i],"i",i
				temp[x][y] = a[startx][i]
				i += 1
				x += 1
			x -= 1
			# i -= 1
		except IndexError:
			print "Error retrieving index: ", i
			# exit()
		# print "first loop done."
		try:
			y += 1
			while y < n-starty:
				# print "x",x,"y",y,"val:",a[startx][i],"i",i
				temp[x][y] = a[startx][i]
				i += 1
				y += 1
			# i -= 1
			y -= 1
		except IndexError:
			print "Error 1 retrieving index: ",i
			# exit()
		# print "second loop done."
		try:
			x -= 1
			while x >= startx:
				# print "x",x,"y",y,"val:",a[startx][i],"i",i
				temp[x][y] = a[startx][i]
				i += 1
				x -= 1
			# i -= 1
			x += 1
		except IndexError:
			print "Error 2 retrieving index: ",i
			# exit()
		# print "Third loop done."
		try:
			y -= 1
			while y >= starty:
				# print "x",x,"y",y,"val:",a[startx][i],"i",i
				temp[x][y] = a[startx][i]
				i += 1
				y -= 1
			y += 1
		except IndexError:
			None
			# print "Error 3 retrieving index: ",i
			# exit()
		startx += 1
		starty += 1

	return temp


I = lambda: map(int, raw_input().split())

def Print(a):
	for s in a:
		for t in s:
			print t,
		print


def Main():
	m, n, k = I()
	a = []
	for i in xrange(m):
		a.append(I())
	# print "a:\n",a
	b = lin(a,m, n)
	# print "b:\n",b
	# print "rotating: ",
	for i in xrange(len(b)):
		# print "b[i]:",b[i],
		b[i] = rotate(b[i], k)

	# print "\nrotated b:\n",b
	c = form(b, m,n)
	# print "c:\n",c
	Print(c)

if __name__ == '__main__':
	Main()

File: hackerrank/practice/matrix-rotation-algo.cpp
#include <iostream>
#include <deque>
#include <typeinfo>

using namespace std;

int min(int n, int m){
	if (n > m){
		return m;
	}else{
		return n;
	}
}

// This function is working properly!
deque<deque<int>> lin(int **a, int m, int n, int k){
	int nol = min(m, n)/2;  // no. of layers in the matrix
	deque<deque<int>> temp;
	int startx = 0, starty = 0, x, y;
	while (startx < nol){
		deque<int> temp2;
		x = startx;
		y = starty;
		while (x < m-startx){
			temp2.push_back(a[x][y]);
			x++;
		}
		x--;
		temp2.pop_back();
		while (y < n-starty){
			temp2.push_back((int)a[x][y]);
			y++;
		}
		y--;
		temp2.pop_back();
		while (x >= startx){
			temp2.push_back((int)a[x][y]);
			x--;
		}
		x++;
		temp2.pop_back();
		while (y >=	starty){
			temp2.push_back((int)a[x][y]);
			y--;
		}
		y++;
		temp2.pop_back();
		startx++;
		starty++;

		// here starts the rotation
		int kk = k%temp2.size();
		while(kk--){
			int val = temp2.back();
			temp2.pop_back();
			temp2.push_front(val);
		}
		temp.push_back(temp2);
	}
	
	return temp;

}

// deque<int> rotate(deque<int> a, int  k){
// 	k %= a.size();
// 	while (k--){
// 		int val = a.pop_back();
// 		a.push_front(val);
// 	}
// 	return a;
// }

int** form(deque<deque<int>> a, int m, int n){
	int **temp, x, y;
	//cout << "Allocating memory for temp..." << endl;
	temp = new int*[m];
	for (int i= 0; i < m; i += 1){
		temp[i] = new int[n];
	}
	for (int i= 0; i < m; i += 1){
		for (int j = 0; j < n; j += 1){
			temp[i][j] = 0;
		}
	}
	//cout << "Done allocating memory for temp, starting to linearize." << endl;
	int startx = 0, starty = 0, i;
	int nol = min(m, n)/2;
	while (startx < nol){
		x = startx;
		y = starty;
		i = 0;
		while (x<m-startx ){ // may the condition x < m be added.
			//if (x == 0 && y == 0) cout <<"temp[0][0]:"<<a[startx][i] << endl;
			temp[x][y] = temp[x][y]==0 ? a[startx][i] : temp[x][y];
			//if (x == 0 && y == 0) cout <<"temp[0][0]:"<<temp[x][y] << endl;

			i += 1;
			x++;
		}
		x--;
		y++;
		while (y<n-starty){
			temp[x][y] = temp[x][y]==0 ? a[startx][i] : temp[x][y];
			i++;
			y++;
		}
		y--;
		x--;
		while (x>=startx){

			temp[x][y] = temp[x][y]==0 ? a[startx][i] : temp[x][y];
			i++;
			x--;
		}
		x++;
		y--;
		while (y>=starty){
			temp[x][y] = temp[x][y]==0 ? a[startx][i] : temp[x][y];
			i++;
			y--;
		}
		y++;
		startx++;
		starty++;
	}
	return temp;

}

void print(deque<int> a){
	for (int i = 0; i < a.size(); i += 1){
		cout << a[i] << " ";
	}
	cout << endl;
}

int main(void){
	int **a, **b,m,n,k;
	
	cin >> m >> n >> k;
	a = new int*[m];
	for (int i = 0; i < m; i += 1) a[i] = new int[n];
	// cout << "Printing the value of a pointer." << endl;
	// for (int i = 0; i < m; i += 1){
	// 	for (int j = 0; j < n; j += 1){
	// 		cout << a[i][j] << " ";
	// 	}
	// 	cout << endl;
	// }
	for (int i = 0; i < m; i ++){
		for (int j = 0; j < n; j += 1){
			cin >> a[i][j];
		}
	}
	//cout << "Done taking input, now linearizing the matrix..."<<endl;
	deque<deque<int>> ad = lin(a, m, n,k);
	// for (int i = 0; i < ad.size(); i += 1){
	// 	print(ad[i]);
	// 	cout << endl;
	// }
	//cout << "Done linearizing and rotating, now reforming the matrix." << endl;
	b = new int*[m];
	for (int i = 0; i < m; i += 1) b[i] = new int[n];
	b = form(ad, m, n);
	//cout << "Done forming the matrix, now printing out."<< endl;
	for (int i = 0;i < m; i += 1){
		for (int j = 0; j < n; j += 1){
			cout << b[i][j] << " ";;
		}
		cout << endl;
	}
}